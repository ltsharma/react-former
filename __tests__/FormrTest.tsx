import React from 'react';
import { render, fireEvent } from 'react-native-testing-library';
import FormrForm from '../example/FormrForm';
import renderer from 'react-test-renderer';
// const createTestProps = (props?: object) => ({
//     ...props
// });

const validations = {
    allRequired: {
        email: { required: true, type: 'email' },
        name: { required: true },
        password: {
            required: true,
            rules:
                '^(?=.*[A-Z].*[A-Z])(?=.*[!@#$&*])(?=.*[0-9].*[0-9])(?=.*[a-z].*[a-z].*[a-z]).{8}$'
        }
    },
    noNameValidation: {
        email: { required: true, type: 'email' },
        password: {
            required: true,
            rules:
                '^(?=.*[A-Z].*[A-Z])(?=.*[!@#$&*])(?=.*[0-9].*[0-9])(?=.*[a-z].*[a-z].*[a-z]).{8}$'
        }
    },
    falseEmailValidation: {
        email: { required: false, type: 'email' },
        name: { required: true },
        password: {
            required: true,
            rules:
                '^(?=.*[A-Z].*[A-Z])(?=.*[!@#$&*])(?=.*[0-9].*[0-9])(?=.*[a-z].*[a-z].*[a-z]).{8}$'
        }
    }
};

const values = {
    withAllRight: {
        email: 'test@testing.com',
        name: 'Tester',
        password: 'TT@eee99'
    },
    emailInvalid: {
        email: 'testtesting',
        name: 'Tester',
        password: 'TT@eee99'
    }
};

describe('Form input test', () => {
    // const props = createTestProps();
    test('changeText', () => {
        const { getByTestId } = render(<FormrForm />);

        const input = getByTestId('nameInput');
        const output = getByTestId('nameInputvalue');
        expect(input).not.toBeNull();

        const tree = renderer.create(<FormrForm />).toJSON();
        expect(tree).toMatchSnapshot();

        const testText = 'Test';
        fireEvent(input, 'onChangeText', testText);
        expect(output.props.children).toBe(testText);
    });
});

describe('Empty submit test', () => {
    test('With all field required', () => {
        const testComponent = (
            <FormrForm validations={validations.allRequired} />
        );
        const { getByTestId } = render(testComponent);

        const button = getByTestId('formSubmit');

        // Input
        const nameInput = getByTestId('nameInput');
        const emailInput = getByTestId('emailInput');
        const passwordInput = getByTestId('passwordInput');
        const submit = getByTestId('submited-check');

        // Test
        expect(button).not.toBeNull();
        expect(nameInput).not.toBeNull();
        expect(emailInput).not.toBeNull();
        expect(passwordInput).not.toBeNull();

        fireEvent.press(button);

        // Values
        expect(nameInput.props.value).toBe('');
        expect(emailInput.props.value).toBe('');
        expect(passwordInput.props.value).toBe('');
        // Valid
        expect(nameInput.props.valid).toBe(false);
        expect(emailInput.props.valid).toBe(false);
        expect(passwordInput.props.valid).toBe(false);
        // Touched
        expect(nameInput.props.touched).toBe(true);
        expect(emailInput.props.touched).toBe(true);
        expect(passwordInput.props.touched).toBe(true);
        // Submit
        expect(submit.props.submited).toBe(false);
    });

    test('Name field without validation', () => {
        const testComponent = (
            <FormrForm validations={validations.noNameValidation} />
        );
        const { getByTestId } = render(testComponent);

        const button = getByTestId('formSubmit');

        // Input
        const nameInput = getByTestId('nameInput');
        const emailInput = getByTestId('emailInput');
        const passwordInput = getByTestId('passwordInput');
        const submit = getByTestId('submited-check');

        // Test
        expect(button).not.toBeNull();
        expect(nameInput).not.toBeNull();
        expect(emailInput).not.toBeNull();
        expect(passwordInput).not.toBeNull();
        fireEvent.press(button);

        // Values
        expect(nameInput.props.value).toBe('');
        expect(emailInput.props.value).toBe('');
        expect(passwordInput.props.value).toBe('');
        // Valid
        expect(nameInput.props.valid).toBeUndefined();
        expect(emailInput.props.valid).toBe(false);
        expect(passwordInput.props.valid).toBe(false);
        // Touched
        expect(nameInput.props.touched).toBe(true);
        expect(emailInput.props.touched).toBe(true);
        expect(passwordInput.props.touched).toBe(true);
        // Submit
        expect(submit.props.submited).toBe(false);
    });
    test('With "email" type set & validation false', () => {
        const testComponent = (
            <FormrForm validations={validations.falseEmailValidation} />
        );
        const { getByTestId } = render(testComponent);

        const button = getByTestId('formSubmit');
        const submit = getByTestId('submited-check');
        // Input
        const nameInput = getByTestId('nameInput');
        const emailInput = getByTestId('emailInput');
        const passwordInput = getByTestId('passwordInput');

        // Test
        expect(button).not.toBeNull();
        expect(nameInput).not.toBeNull();
        expect(emailInput).not.toBeNull();
        expect(passwordInput).not.toBeNull();
        fireEvent.press(button);

        // Values
        expect(nameInput.props.value).toBe('');
        expect(emailInput.props.value).toBe('');
        expect(passwordInput.props.value).toBe('');
        // Valid
        expect(nameInput.props.valid).toBe(false);
        expect(emailInput.props.valid).toBe(false);
        expect(passwordInput.props.valid).toBe(false);
        // Touched
        expect(nameInput.props.touched).toBe(true);
        expect(emailInput.props.touched).toBe(true);
        expect(passwordInput.props.touched).toBe(true);
        // Submit
        expect(submit.props.submited).toBe(false);
    });
});

describe('Submit with value test', () => {
    test('Submit success with all valid', () => {
        const testComponent = (
            <FormrForm validations={validations.allRequired} />
        );
        const { getByTestId } = render(testComponent);

        const button = getByTestId('formSubmit');
        const submit = getByTestId('submited-check');
        // Input
        const nameInput = getByTestId('nameInput');
        const emailInput = getByTestId('emailInput');
        const passwordInput = getByTestId('passwordInput');

        // Test
        expect(button).not.toBeNull();
        expect(nameInput).not.toBeNull();
        expect(emailInput).not.toBeNull();
        expect(passwordInput).not.toBeNull();

        fireEvent(nameInput, 'onChangeText', values.withAllRight.name);
        fireEvent(nameInput, 'onBlur');
        fireEvent(emailInput, 'onChangeText', values.withAllRight.email);
        fireEvent(emailInput, 'onBlur');
        fireEvent(passwordInput, 'onChangeText', values.withAllRight.password);
        fireEvent(passwordInput, 'onBlur');
        fireEvent.press(button);
        // Values
        expect(nameInput.props.value).toBe(values.withAllRight.name);
        expect(emailInput.props.value).toBe(values.withAllRight.email);
        expect(passwordInput.props.value).toBe(values.withAllRight.password);
        // Valid
        expect(nameInput.props.valid).toBe(true);
        expect(emailInput.props.valid).toBe(true);
        expect(passwordInput.props.valid).toBe(true);
        // Touched
        expect(nameInput.props.touched).toBe(true);
        expect(emailInput.props.touched).toBe(true);
        expect(passwordInput.props.touched).toBe(true);
        // Submit
        expect(submit.props.submited).toBe(true);
    });
    test('Submit fail with invalid email', () => {
        const testComponent = (
            <FormrForm validations={validations.allRequired} />
        );
        const { getByTestId } = render(testComponent);

        const button = getByTestId('formSubmit');
        const submit = getByTestId('submited-check');
        // Input
        const nameInput = getByTestId('nameInput');
        const emailInput = getByTestId('emailInput');
        const passwordInput = getByTestId('passwordInput');

        // Test
        expect(button).not.toBeNull();
        expect(nameInput).not.toBeNull();
        expect(emailInput).not.toBeNull();
        expect(passwordInput).not.toBeNull();

        fireEvent(nameInput, 'onChangeText', values.emailInvalid.name);
        fireEvent(nameInput, 'onBlur');
        fireEvent(emailInput, 'onChangeText', values.emailInvalid.email);
        fireEvent(emailInput, 'onBlur');
        fireEvent(passwordInput, 'onChangeText', values.emailInvalid.password);
        fireEvent(passwordInput, 'onBlur');
        fireEvent.press(button);
        // Values
        expect(nameInput.props.value).toBe(values.emailInvalid.name);
        expect(emailInput.props.value).toBe(values.emailInvalid.email);
        expect(passwordInput.props.value).toBe(values.emailInvalid.password);
        // Valid
        expect(nameInput.props.valid).toBe(true);
        expect(emailInput.props.valid).toBe(false);
        expect(passwordInput.props.valid).toBe(true);
        // Touched
        expect(nameInput.props.touched).toBe(true);
        expect(emailInput.props.touched).toBe(true);
        expect(passwordInput.props.touched).toBe(true);
        // Submit
        expect(submit.props.submited).toBe(false);
    });

    test('Submit success with invalid email & email is not required', () => {
        const testComponent = (
            <FormrForm validations={validations.falseEmailValidation} />
        );
        const { getByTestId } = render(testComponent);

        const button = getByTestId('formSubmit');
        const submit = getByTestId('submited-check');
        // Input
        const nameInput = getByTestId('nameInput');
        const emailInput = getByTestId('emailInput');
        const passwordInput = getByTestId('passwordInput');

        // Test
        expect(button).not.toBeNull();
        expect(nameInput).not.toBeNull();
        expect(emailInput).not.toBeNull();
        expect(passwordInput).not.toBeNull();

        fireEvent(nameInput, 'onChangeText', values.emailInvalid.name);
        fireEvent(nameInput, 'onBlur');
        fireEvent(emailInput, 'onChangeText', values.emailInvalid.email);
        fireEvent(emailInput, 'onBlur');
        fireEvent(passwordInput, 'onChangeText', values.emailInvalid.password);
        fireEvent(passwordInput, 'onBlur');
        fireEvent.press(button);
        // Values
        expect(nameInput.props.value).toBe(values.emailInvalid.name);
        expect(emailInput.props.value).toBe(values.emailInvalid.email);
        expect(passwordInput.props.value).toBe(values.emailInvalid.password);
        // Valid
        expect(nameInput.props.valid).toBe(true);
        expect(emailInput.props.valid).toBe(false);
        expect(passwordInput.props.valid).toBe(true);
        // Touched
        expect(nameInput.props.touched).toBe(true);
        expect(emailInput.props.touched).toBe(true);
        expect(passwordInput.props.touched).toBe(true);
        // Submit
        expect(submit.props.submited).toBe(true);
    });
});
