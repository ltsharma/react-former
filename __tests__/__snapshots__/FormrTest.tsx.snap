// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form input test changeText 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView
    style={
      Object {
        "backgroundColor": "#fff",
      }
    }
  >
    <View>
      <Text>
        Former
      </Text>
      <View
        style={
          Object {
            "flex": 1,
            "marginTop": 50,
          }
        }
      >
        <Text>
          Welcome
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#EAEAEA",
              "marginTop": 16,
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          <Text>
            <Text
              style={
                Object {
                  "fontSize": 16,
                  "fontWeight": "600",
                }
              }
            >
              Email:
            </Text>
             
            Required input with email type
          </Text>
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            style={
              Object {
                "borderColor": "black",
                "borderWidth": 1,
                "height": 46,
                "marginBottom": 16,
                "marginTop": 8,
                "padding": 8,
              }
            }
            testID="emailInput"
            touched={false}
            value=""
          />
        </View>
        <Text
          testID="emailInputvalue"
        />
        <Text
          testID="emailInputvalid"
        >
          false
        </Text>
        <Text
          testID="emailtouched"
        >
          false
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#EAEAEA",
              "marginTop": 16,
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          <Text>
            <Text
              style={
                Object {
                  "fontSize": 16,
                  "fontWeight": "600",
                }
              }
            >
              Name:
            </Text>
             
            Required input with no rules or type set
          </Text>
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            style={
              Object {
                "borderColor": "black",
                "borderWidth": 1,
                "height": 46,
                "marginBottom": 16,
                "marginTop": 8,
                "padding": 8,
              }
            }
            testID="nameInput"
            touched={false}
            value=""
          />
        </View>
        <Text
          testID="nameInputvalue"
        />
        <Text
          testID="nameInputvalid"
        >
          false
        </Text>
        <Text
          testID="nameInputtouched"
        >
          false
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#EAEAEA",
              "marginTop": 16,
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          <Text>
            <Text
              style={
                Object {
                  "fontSize": 16,
                  "fontWeight": "600",
                }
              }
            >
              Password:
            </Text>
             
            Required with rules set
             
          </Text>
          <Text>
            Rules (regex): has two uppercase letters. has one special case letter. has two digits. has three lowercase letters. string is of length 8.
          </Text>
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            style={
              Object {
                "borderColor": "black",
                "borderWidth": 1,
                "height": 46,
                "marginBottom": 16,
                "marginTop": 8,
                "padding": 8,
              }
            }
            testID="passwordInput"
            touched={false}
            value=""
          />
          <Text
            testID="passwordInputvalue"
          />
          <Text
            testID="passwordInputvalid"
          >
            false
          </Text>
          <Text
            testID="passwordInputtouched"
          >
            false
          </Text>
        </View>
        <View
          accessibilityLabel="Learn more about this purple button"
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="formSubmit"
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#ff5687",
                  },
                ]
              }
            >
              Submit
            </Text>
          </View>
        </View>
        <View>
          <Text>
            Submitted Values:
          </Text>
          <Text>
            {}
          </Text>
        </View>
        <View
          submited={false}
          testID="submited-check"
        />
        <View>
          <Text>
            On Change values:
          </Text>
          <Text>
            {}
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
